src/Embedders/Word2Vec.php

```diff
@@ -0,0 +1,922 @@
+            throw new InvalidArgumentException("Dimensions must be greater than 4, $dimensions given.");
+        }
+
+        if ($sampleRate < 0.0) {
+            throw new InvalidArgumentException("Sample rate must be 0 or greater, $sampleRate given.");
+        }
+
+        if ($alpha < 0.0) {
+            throw new InvalidArgumentException("Alpha must be greater than 0, $alpha given.");
+        }
+
+        if ($epochs < 1) {
+            throw new InvalidArgumentException("Number of epochs must be greater than 0, $epochs given.");
+        }
+
+        if ($minCount < 1) {
+            throw new InvalidArgumentException("Minimum word count must be greater than 0, $minCount given.");
+        }
+
+        $this->layer = $layer;
+        $this->window = $window;
+        $this->dimensions = $dimensions;
+        $this->sampleRate = $sampleRate;
+        $this->alpha = $alpha;
+        $this->epochs = $epochs;
+        $this->minCount = $minCount;
+        $this->negLabels = Vector::quick([1, 0]);
+        $this->sigmoid = new Sigmoid();
+    }
+
+    /**
+     * Return the data types that this estimator is compatible with.
+     *
+     * @return \Rubix\ML\DataType[]
+     */
+    public function compatibility() : array
+    {
+        return [
+            DataType::categorical(),
+        ];
+    }
+
+    /**
+     * Return the settings of the hyper-parameters in an associative array.
+     *
+     * @return mixed[]
+     */
+    public function params() : array
+    {
+        return [
+            'layer' => $this->layer,
+            'window' => $this->window,
+            'dimensions' => $this->dimensions,
+            'sample_rate' => $this->sampleRate,
+            'alpha' => $this->alpha,
+            'epochs' => $this->epochs,
+            'min_count' => $this->minCount,
+        ];
+    }
+
+    /**
+     * Has the learner been trained?
+     *
+     * @return bool
+     */
+    public function fitted() : bool
+    {
+        return !empty($this->vectors);
+    }
+
+    /**
+     * Iterating through a range of the specific epoch count and updating all respective word vectors.
+     * 
+     * @param \Rubix\ML\Datasets\Dataset $dataset
+     * @throws \InvalidArgumentException
+     */
+    public function fit(Dataset $dataset) : void
+    {
+        DatasetIsNotEmpty::check($dataset);
+
+        $sentences = $dataset->column(0);
+
+        $this->preprocess($sentences);
+        $this->prepareWeights();
+
+        $startAlpha = $this->alpha;
+
+        for ($i = 0; $i < $this->epochs; ++$i) {
+            $this->alpha = $startAlpha - (($startAlpha - self::MIN_ALPHA) * ($i) / $this->epochs);
+
+            $this->trainEpochSg();
+        }
+
+        $this->generateL2Norm();
+    }
+
+    /**
+     * Embed a high dimensional dataset into a lower dimensional one.
+     *
+     * @param array[] $samples
+     * @throws \RuntimeException
+     */
+    public function transform(array &$samples) : void
+    {
```

Comment:
You are not mutating the `$samples` array passed by reference here, instead you are creating an `$embeddedDataset` array that doesn't get assigned to anything outside the method context.

Each sample in the samples array should have its categorical/string feature columns replaced with its word embedding